### PR TITLE
Added staff role on ccx to all the staff members of master course.

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
### Background
From issue https://github.com/mitocw/edx-platform/issues/126

### What is done in this PR
**Studio Updates:** None.

**LMS Updates:** 
- Now when a coach creates a new ccx, I am assigning staff role on this ccx course to all staff of master course.
- Now staff of master course has staff rights on master course as well as ccx courses.
- This is for course staff to see what ccx coaches are doing with course.
- Move code that is not rendering view to utility.

@pdpinch @giocalitri @pwilkins 

- Staff on Master course
screen shot 2015-11-30 at 5 12 54 pm

- Staff on CCX
screen shot 2015-11-30 at 5 13 04 pm

### CODE REVIEW

File: common/djangoapps/student/roles.py [hunk 1]
```diff
@@ -47,9 +47,11 @@ def has_role(self, role, course_id, org):
         """
         Return whether this RoleCache contains a role with the specified role, course_id, and org
         """
+        # Convert course keys to unicode string before comparison. this is because some objects
+        # like ccx locator dont have overridden __eq__. Thats why when we compare them always return false.
         return any(
             access_role.role == role and
-            access_role.course_id == course_id and
+            unicode(access_role.course_id) == unicode(course_id) and
```
Comment (amir-qayyum-khan) [inline common/djangoapps/student/roles.py]: Because code was not able to compare ccxLocator objects though it was able to compare courseLocator.
In sum when this line was comparing ccx keys it always give false even both key are match. Thats why I am deserialising keys and comparing them as strings.

second option is block this PR and create small PR against 
https://github.com/edx/ccx-keys/blob/master/ccx_keys/locator.py

and override eq method. Because we dont have this method thats why when i compare 2 ccx locator objects it always returns false

After that resume this PR
http://stackoverflow.com/questions/1227121/compare-object-instances-for-equality-by-their-attributes-in-python

Comment (pdpinch) [inline common/djangoapps/student/roles.py]: I don't see a corresponding __eq__ method in https://github.com/edx/opaque-keys/blob/master/opaque_keys/edx/locator.py -- how is this dealt with elsewhere in edX?

File: lms/djangoapps/ccx/views.py [hunk 2]
```diff
@@ -560,6 +564,28 @@ def ccx_course(ccx_locator):
     yield course
 
 
+def add_master_course_staff_to_ccx(course, ccx_key, display_name):
+    """
+    Assign staff of master course, staff role on ccx.
+    :param course: Master course
```
Comment (pdpinch) [inline lms/djangoapps/ccx/views.py]: is this a master course key?

File: lms/djangoapps/instructor/access.py [hunk 3]
```diff
@@ -88,7 +88,10 @@ def _change_access(course, user, level, action):
                 email_students=True,
                 email_params=email_params,
             )
-        role.add_users(user)
+        print "role.has_user(user): ", role.has_user(user)
```
Comment (pdpinch) [inline lms/djangoapps/instructor/access.py]: debugging code?

File: lms/djangoapps/instructor/access.py [hunk 4]
```diff
@@ -88,7 +88,10 @@ def _change_access(course, user, level, action):
...
+        if not role.has_user(user):
+            # If user does not have this role, only then assign him.
+            role.add_users(user)
```
Comment (pdpinch) [inline lms/djangoapps/instructor/access.py]: why is this necessary?

File: lms/djangoapps/ccx/views.py (hunk 2 above)
Comment (amir-qayyum-khan) [inline lms/djangoapps/ccx/views.py]: No it is master course object

**Comment by amir-qayyum-khan** [comment]
@pdpinch fixed issues

**Comment by amir-qayyum-khan** [comment]
@pdpinch should i create this on upstream?

**Comment by pdpinch** [comment]
Ok. Can you add "wip" to the title, until someone from our team can give it a thumbs up? 

----------------
I am mobile.

File: lms/djangoapps/ccx/views.py [hunk 5]
```diff
@@ -572,6 +576,28 @@ def ccx_course(ccx_locator):
     yield course
 
 
+def add_master_course_staff_to_ccx(course, ccx_key, display_name):
+    """
+    Assign staff of master course, staff role on ccx.
+    :param course: Master course instance
+    :param ccx_key: CCX course key
+    :param display_name: ccx display name for email
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: I'm not sure this is the right or current way they want the docstrings.

File: lms/djangoapps/ccx/views.py [hunk 6]
```diff
@@ -572,6 +576,28 @@ def ccx_course(ccx_locator):
     yield course
 
 
+def add_master_course_staff_to_ccx(course, ccx_key, display_name):
+    """
+    Assign staff of master course, staff role on ccx.
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: to whom?

File: lms/djangoapps/ccx/views.py [hunk 7]
```diff
@@ -572,6 +576,28 @@ def ccx_course(ccx_locator):
...
+    :param course: Master course instance
+    :param ccx_key: CCX course key
+    :param display_name: ccx display name for email
+    """
+    list_staff = list_with_level(course, 'staff')
+    with ccx_course(ccx_key) as course:
+        email_params = get_email_params(course, auto_enroll=True, course_key=ccx_key, display_name=display_name)
+        for staff in list_staff:
+            allow_access(course, staff, 'staff')
+            # Enroll the coach in the course
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: you are enrolling not only the coach, but any other staff member, if I understood well

File: lms/djangoapps/ccx/views.py [hunk 8]
```diff
@@ -572,6 +576,28 @@ def ccx_course(ccx_locator):
...
+            enroll_email(
+                course_id=ccx_key,
+                student_email=staff.email,
+                auto_enroll=True,
+                email_students=True,
+                email_params=email_params,
+            )
+
+
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: I think this function looks fine.

Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: I think this function looks fine, but what I meant in my comment in the story was to put this function in an external module ( `api.py` ?) so I can import it easily.

**Comment by amir-qayyum-khan** [comment]
Done @giocalitri

File: lms/djangoapps/ccx/api.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    get_email_params,
+)
+
+
+def add_master_course_staff_to_ccx(course, ccx_key, display_name):
+    """
+    Added staff role on ccx to all the staff members of master course. 
+    :param course: Master course instance
+    :param ccx_key: CCX course key
+    :param display_name: ccx display name for email
```
Comment (giocalitri) [inline lms/djangoapps/ccx/api.py]: Sorry to insist on this, but I think the way they want docstrings now is something like this:

```
Args:
    course (<the object type here>): Master course instance
    ccx_key (<the object type here>): CCX course key
    ...

Returns:
    None  # Not sure this is actually necessary
```

File: lms/djangoapps/ccx/api.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    """
+    # To avoid circular import
+    from views import ccx_course
```
Comment (giocalitri) [inline lms/djangoapps/ccx/api.py]: you should move this outside the `views` file as well

Comment (giocalitri) [inline lms/djangoapps/ccx/api.py]: And if there are tests related, we should also move the tests to another file

**Comment by giocalitri** [comment]
I know you put the test in the `test_views.py`, but I feel like we need another one in `test_apis.py`

File: lms/djangoapps/ccx/api.py (hunk 9 above)
Comment (giocalitri) [inline lms/djangoapps/ccx/api.py]: Actually I have seen both, so I am not sure

**Comment by amir-qayyum-khan** [comment]
@giocalitri Done with moving code and creating test file.
Removed ``api.py`` because the function i made is not api but utility function.
So i moved it in utility.
Updated comment as well.

CC @pdpinch

File: lms/djangoapps/ccx/tests/test_utils.py [hunk 11]
```diff
@@ -3,17 +3,28 @@
...
 from xmodule.modulestore.tests.factories import CourseFactory
+from xmodule.modulestore.django import modulestore
+
+from student.roles import CourseCcxCoachRole, CourseStaffRole
+
+from instructor.access import list_with_level, allow_access
 
 from ccx_keys.locator import CCXLocator
+from lms.djangoapps.ccx.overrides import override_field_for_ccx
+from ..utils import add_master_course_staff_to_ccx, ccx_course
```
Comment (giocalitri) [inline lms/djangoapps/ccx/tests/test_utils.py]: can please you use the full path for this? Even if it is correct I would avoid using this kind of imports unless the module is in the same directory of the file that imports it

**Comment by giocalitri** [comment]
I think you did a great job about moving all the functions in the new module.

**Comment by amir-qayyum-khan** [comment]
Made upstream PR https://github.com/edx/edx-platform/pull/10877
